cpp/src/arrow/compute/exec/aggregate_node.cc

```diff
@@ -81,10 +152,22 @@ class ScalarAggregateNode : public ExecNode {
 
     const auto& aggregate_options = checked_cast<const AggregateNodeOptions&>(options);
     auto aggregates = aggregate_options.aggregates;
+    const auto& segment_keys = aggregate_options.segment_keys;
 
     const auto& input_schema = *inputs[0]->output_schema();
     auto exec_ctx = plan->exec_context();
 
+    std::vector<int> segment_field_ids(segment_keys.size());
+    std::vector<TypeHolder> segment_key_types(segment_keys.size());
+    for (size_t i = 0; i < segment_keys.size(); i++) {
+      ARROW_ASSIGN_OR_RAISE(auto match, segment_keys[i].FindOne(input_schema));
+      segment_field_ids[i] = match[0];
```

Comment:
```suggestion
      if (match.indices().size() > 1) {
        return Status::Invalid("Nested references cannot be used as segment ids");
      }
      segment_field_ids[i] = match[0];
```

And then maybe make a JIRA to support this at some point (though maybe not, our story on nested columns is pretty muddled at the moment).  Or we could find some way to extract a nested path from an exec batch to support it now.

Reply:
https://issues.apache.org/jira/browse/ARROW-18369